Integrating Codegen framework with executor flow

karthijrk opened the pull request:
In this PR, we integrate codegen framework with executor flow and shows how llvm calls slot_deform_tuple as external function. 
Acceptance Test for this PR: 
```
1. Init and start gpdb
2. Set codegen = on using gpconfig
E.g gpconfig -c codegen -v true
3. gpstop -ra
4. Select * from t; (Any table)
```
This should print elog message INFO:  Calling wrapped function: slot_deform_tuple

hardikar commented on src/backend/codegen/CMakeLists.txt:
```diff
@@ -113,7 +127,7 @@ include_directories(${CLANG_INCLUDE_DIRS})
 add_library(gpcodegen SHARED
             utils/clang_compiler.cc
-            utils/code_generator.cc
+            utils/codegen_utils.cc
```
These may be codegen utils. But more specifically - this class is a wrapper around llvm - it's IRBuilder, IROptimizer etc. I bring this up because both clang_compiler and code_generator are codegenutils which is why they have that namespace. This name should be more instructive in my opinion.

foyzur commented on src/backend/access/common/heaptuple.c:
```diff
@@ -1241,8 +1244,11 @@ _slot_getsomeattrs(TupleTableSlot *slot, int attnum)
 	attno = Min(attno, attnum);
 
+#ifdef USE_CODEGEN
+	slot->slot_deform_tuple_gen_info.slot_deform_tuple_fn(slot, attno);
```
Can we relocate all these ifdef in codegen_wrapper.h? E.g.,

```
#ifdef USE_CODEGEN
#define call_slot_deform_tuple(slot, attno) slot->slot_deform_tuple_gen_info.slot_deform_tuple_fn(slot, attno);
#else
#define call_slot_deform_tuple(slot, attno) slot_deform_tuple(slot, attno)
#endif

```

karthijrk replied:
Yes . Good suggestion. I will fix that.

armenatzoglou commented:
I suggest us to open a PR with the renaming changes (e.g., code_generator -> codegen_utils) only. 18 out of 39 files contain only this kind of changes. The rest of the files have also many renaming changes. I believe that this will the reviewers of the second PR, which will be on the framework only.

hardikar commented on src/backend/codegen/CMakeLists.txt:
```diff
@@ -162,26 +183,121 @@ target_link_libraries(
+    COMMAND ${CMAKE_COMMAND} .
+    COMMAND ${CMAKE_MAKE_PROGRAM} check
+)
+include_directories(${TOP_SRC_DIR}/src/test/unit/cmockery)
```
NOTE: use target_include_directories(TARGET ${DIR}) instead of including cmockery headers for other targets here.

karthijrk replied:
Cool, Let me make changes.

hardikar commented on src/backend/codegen/codegen_interface.cc:
```diff
@@ -0,0 +1,21 @@
+
+using gpcodegen::CodeGenInterface;
+
+unsigned CodeGenInterface::unique_counter_ = 0;
```
Missing comments.

karthijrk replied:
Comments for this variable is in header file. Do you still see a need to add here ?

armenatzoglou commented on src/include/executor/tuptable.h:
```diff
@@ -140,6 +148,8 @@ typedef struct TupleTableSlot
     /* System attributes */
     Oid         tts_tableOid;
+
+    SlotDeformTupleCodeGenInfo slot_deform_tuple_gen_info;
```
I have two issues here:
- Even if we do not codegen, we allocate memory sizeof(SlotDeformTupleCodeGenInfo). 
- In the future, we will have numerous codegened functions, whose function pointer can be reached through slot. Are we going to allocate memory for all structs of the new functions here as well?

armenatzoglou replied:
I think that we need to have a pointer to another struct that has pointers to all xxxxCodeGenInfo structs. In that case, we will only have a pointer (e.g., CodeGenFunctions *) in TupleTableSlot.

```
typedef struct CodeGenFunctions{
     SlotDeformTupleCodeGenInfo *slot_deform_tuple_gen_info;
} CodeGenFunctions;

struct TupleTableSlot{
    ....
    ....
    CodeGenFunctions *codegened_funcs;
}
```

hardikar replied:
Another place or #define USE_CODEGEN?

foyzur replied:
May be wrap with ifdef. We need inlined version for performance.

karthijrk replied:
Let me wrap it with #ifdef USE_CODEGEN

armenatzoglou commented on src/include/executor/tuptable.h:
```diff
@@ -110,6 +111,13 @@
+{
+	void* code_generator;
+	SlotDeformTupleFn slot_deform_tuple_fn;
+} SlotDeformTupleCodeGenInfo;
```
Add comments/descriptions.

karthijrk replied:
Sure, let me add some more comments on that.

hardikar commented on src/backend/codegen/include/codegen/codegen_interface.h:
```diff
@@ -0,0 +1,109 @@
+ */
+
+// Forward declaration
+class CodeGenUtils;
```
I would rather it be name as something else - related to LLVM or IR. but if we're going to keep this as it, please rename this to CodegenUtils to match the file: codegen_utils.cc

hardikar commented on src/backend/codegen/codegen_manager.cc:
```diff
@@ -0,0 +1,90 @@
+using gpcodegen::CodeGenManager;
+
+CodeGenManager::CodeGenManager() {
+  codegen_utils_.reset(new gpcodegen::CodeGenUtils("test_module"));
```
Why do we initialize with  this "test_module"?

foyzur replied:
This is NOT a test_module. We should give it a more meaningful name. One manager per operator. So, perhaps we can pass the operator's name and derive manager's name from that?

karthijrk replied:
Let me change it to string as an argument in constructor and let operator to pass its name for that.

foyzur commented on src/backend/codegen/CMakeLists.txt:
```diff
@@ -113,7 +127,7 @@ include_directories(${CLANG_INCLUDE_DIRS})
             utils/clang_compiler.cc
-            utils/code_generator.cc
+            utils/codegen_utils.cc
             ${codegen_tmpfile_sources})
```
what is this tmpfile source?

hardikar replied:
This came from Balerion - I believe its' used for debugging purposes. I'm not fully sure.

karthijrk replied:
I believe this dump c++ code for debugging to some temporary file

foyzur commented on src/backend/codegen/codegen_interface.cc:
```diff
@@ -0,0 +1,21 @@
+
+unsigned CodeGenInterface::unique_counter_ = 0;
+
+std::string CodeGenInterface::GenerateUniqueName(
```
I wish we could find a better place for this.

armenatzoglou commented on src/backend/codegen/codegen_wrapper.cc:
```diff
@@ -0,0 +1,85 @@
+
+  SlotDeformTupleCodeGen* generator = new SlotDeformTupleCodeGen(
+      slot, regular_func_ptr, ptr_to_chosen_func_ptr);
+  assert(nullptr != manager);
```
Do you mean assert(nullptr != generator) ? i.e., generator instead of manager

karthijrk replied:
I think that line is redundant. Initially we didn't have 
`if (nullptr == manager)` at line # 71. So we had that assert for manager. Now given we have if check already , I don't see a need for assert for manager. Let me remove that. Good catch :)

foyzur commented on src/backend/codegen/codegen_manager.cc:
```diff
@@ -0,0 +1,90 @@
+bool CodeGenManager::PrepareGeneratedFunctions() {
+  // Call CodeGenUtils to compile entire module
+  bool compilation_status = codegen_utils_->PrepareForExecution(
+      gpcodegen::CodeGenUtils::OptimizationLevel::kNone, true);
```
We need customization of optimization level in manager, depending on operator, perhaps.

karthijrk replied:
I prefer to skip it now for the first version. Let me know if you have different opinion

foyzur commented on src/backend/codegen/codegen_manager.cc:
```diff
@@ -0,0 +1,90 @@
+  // the pointer so compiled function get called
+  gpcodegen::CodeGenUtils* codegen_utils = codegen_utils_.get();
+  for (auto& generator : enrolled_code_generators_) {
+    generator->SetToGenerated(codegen_utils);
```
We should do something about the return value from SetToGenerated.

karthijrk replied:
Yes, I changed return value of PrepareForExecution from bool to unsigned int

foyzur commented on src/backend/codegen/codegen_wrapper.cc:
```diff
@@ -0,0 +1,85 @@
+
+// Perform global set-up tasks for code generation. Returns 0 on
+// success, nonzero on error.
+int InitCodeGen() {
```
Let's change it to bool status. So, return true on success and false on failure. Note, this is a reversal of semantics.

hardikar replied:
Or change the rest of the functions to return an integer (consistent with postgres?). That way we can return more information with this return code: eg CodeGeneratorManager_PrepareGeneratedFunction -> the number that failed?

karthijrk replied:
I prefer to change all bool to int like @hardikar  suggested. @foyzur  do you agree.

foyzur commented on src/backend/codegen/codegen_wrapper.cc:
```diff
@@ -0,0 +1,85 @@
+}
+
+void* SlotDeformTupleCodeGen_Enroll(
+    TupleTableSlot* slot,
```
Refactor this method into an utility method with variadic template that takes the code generator sub-class type and then can create an instance and enroll it with manager.

karthijrk replied:
Yes that's possible. That's a good idea. I like it. Since some other class may need more than one this pointer, I will move the this ptr as last argument.

foyzur commented on src/backend/codegen/codegen_wrapper.cc:
```diff
@@ -0,0 +1,85 @@
+  CodeGenManager* manager = static_cast<CodeGenManager*>(
+      GetActiveCodeGeneratorManager());
+
+  if (nullptr == manager) {
```
Handling nullptr manager looks bad specialization. May be a default manager is a better design, where the default manager is a no-op other than setting to static version. That way we also have better debugging option as we don't loose the enrollment information.

karthijrk replied:
I prefer to keep it simple with null check. Can we have another story to enhance this ?

hardikar commented on src/backend/codegen/codegen_manager.cc:
```diff
@@ -0,0 +1,90 @@
+
+size_t CodeGenManager::GenerateCode() {
+  size_t success_count = 0;
+  for (auto& generator : enrolled_code_generators_) {
```
Replace auto with the type please. I would rather auto be used when it's cumbersome and very obvious or in templates.

karthijrk replied:
Sure, I will change it.

hardikar commented on src/backend/codegen/include/codegen/base_codegen.h:
```diff
@@ -0,0 +1,158 @@
+ * @tparam FuncPtrType Function type for regular / generated functions.
+ **/
+template<class FuncPtrType>
+class BaseCodeGen: public CodeGenInterface {
```
BaseCodeGenerator?

karthijrk replied:
When I say Codegenerator it sounds like it is generating the code but actually it does more than that. So I prefer codegen but open to change.

armenatzoglou commented:
I believe that we need a document that describes the framework and gives an example using slot_deform_tuples. Also, in this document we can have a section "Codegening a function", which describes the steps that a developer needs to follow to codegen a function. Thanks to this framework, the steps can be clearly defined. This will be helpful for us, the PR, and open source community since they will be able to contribute as well :)

I can work on that. @vraghavan78 Do you agree? We can create a story on that.

foyzur commented on src/backend/codegen/include/codegen/base_codegen.h:
```diff
@@ -0,0 +1,158 @@
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+  }
+
+  bool GenerateCode(gpcodegen::CodeGenUtils* codegen_utils) final {
```
Comments on all the public functions.

karthijrk replied:
This is overriding virtual function in interface. We have necessary comments in the base class

hardikar replied:
Still I believe we have the comments multiple times. There might may be something interesting to say here :)

foyzur commented on src/backend/codegen/include/codegen/base_codegen.h:
```diff
@@ -0,0 +1,158 @@
+    ptr_to_chosen_func_ptr_(ptr_to_chosen_func_ptr),
+    is_generated_(false) {
+    // Initialize the caller to use regular version of generated function.
+    SetToRegular(regular_func_ptr, ptr_to_chosen_func_ptr);
```
As I mentioned above, let's introduce a default manager, and then we can make all the SetToRegular and SetToGenerated protected (or even private if we can swap the pointers in the base code gen). Given that both the regular_func_ptr_ and ptr_to_chosen_func_ptr_ are in this class, it makes sense to make the methods that operate on these as private.

foyzur commented on src/backend/codegen/tests/codegen_framework_unittest.cc:
```diff
@@ -0,0 +1,324 @@
+  ASSERT_TRUE(SumFuncRegular == failed_func_ptr);
+}
+
+/*TEST_F(CodeGenManagerTest, PrepareGeneratedFunctionsCompilationErrorTest) {
```
Commented tests should not be in PR

karthijrk replied:
Good catch. Let me remove that.

foyzur commented on src/backend/executor/execTuples.c:
```diff
@@ -359,6 +361,12 @@ ExecSetSlotDescriptor(TupleTableSlot *slot,		/* slot to change */
 
 		MemoryContextSwitchTo(oldcontext);
 	}
+	slot->slot_deform_tuple_gen_info.code_generator = SlotDeformTupleCodeGen_Enroll(slot,
```
Enrollment should also be in the ifdef. Or if we decide to put ifdef in wrapper, we need to do so uniformly.

karthijrk replied:
Let me move this to codegen_wrapper.

hardikar commented on src/backend/codegen/CMakeLists.txt:
```diff
@@ -162,26 +183,121 @@ target_link_libraries(
+
+# Add CMockery tests
+if(EXISTS ${TXT_OBJFILE})
+    add_cmockery_gtest(codegen_framework_unittest 
```
We should rename this to codegen_framework_unittest.t so that git ignore rules apply and these files don't clutter the git status

karthijrk replied:
oh! Let me modify that

hardikar commented on src/backend/codegen/include/codegen/utils/utility.h:
```diff
@@ -32,7 +33,7 @@ namespace gpcodegen {
  *         position was beyond the end of function's arguments.
  **/
-llvm::Argument* ArgumentByPosition(llvm::Function* function,
+static llvm::Argument* ArgumentByPosition(llvm::Function* function,
```
Why did you make this a static function? Won't this be accessed from our code generators?

karthijrk replied:
It's in header file. So if it is not static, then it gives error as "multiple definition of symbol" as this header file is included from different place.

gcaragea replied:
Shouldn't it be "extern" then?
EDIT: Nevermind, I didn't see it's a full definition here.

hardikar commented on src/backend/codegen/include/codegen/utils/codegen_utils.h:
```diff
@@ -398,13 +404,29 @@ class CodeGenerator {
+     *         function_name, or NULL if the function is not found or couldn't
+     *         be compiled.
+     **/
+    template <typename FunctionType>
```
Check indentation.

karthijrk replied:
Good catch :) I will fix it.

hardikar commented on src/backend/codegen/include/codegen/utils/codegen_utils.h:
```diff
@@ -940,21 +963,80 @@ llvm::Value* CodeGenerator::GetPointerToMemberImpl(
+template<typename ReturnType, typename... ArgumentTypes>
+class FunctionTypeUnpacker<ReturnType(*)(ArgumentTypes...)> {
+ public:
+  using R = ReturnType;
```
not used?

karthijrk replied:
It is used by GetFunctionPointerTypeDef and CreateFunctionTypeDef. around line # 1013

hardikar commented on src/backend/codegen/include/codegen/utils/codegen_utils.h:
```diff
@@ -249,8 +249,14 @@ class CodeGenerator {
   }
 
+  template <typename FuncType>
+  llvm::Function* CreateFunctionTypeDef(
```
1. Missing comments
2. Maybe use Typedef since that's one word in C land 
3. Why do we need this - are we not able to infer the function type automatically based on the return type?

karthijrk replied:
Like we discussed, I will make default GetFunctionPointer to take typedef function type and change existing GetFuctionPointer to GetFunctionPointerImp private function

gcaragea commented on src/backend/codegen/include/codegen/base_codegen.h:
```diff
@@ -0,0 +1,158 @@
+   **/
+  static bool SetToRegular(FuncPtrType regular_func_ptr,
+                           FuncPtrType* ptr_to_chosen_func_ptr) {
+    *ptr_to_chosen_func_ptr = regular_func_ptr;
```
Not sure what we decided in terms of assertions for C++ code. Normally I'd suggest some assertions here for the arguments.

karthijrk replied:
Good point. I believe we can use c++'s assert from cassert. I will try and update it.

gcaragea commented on src/backend/codegen/include/codegen/base_codegen.h:
```diff
@@ -0,0 +1,158 @@
+   * @brief Constructor
+   *
+   * @param orig_func_name         Original function name.
+   * @param regular_func_ptr       Regular version of the generated function.
```
Can we call it something else instead of generated function? Maybe "target" function?

karthijrk replied:
@gcaragea  Are you talking about this comment "Regular version of the generated function". This will change to something like "Regular version of the target function" ? I like this to call as target function :)

gcaragea replied:
Yes.. Generated function doesn't sound right - especially since this is the non-generated function :)

karthijrk replied:
Sounds good. I like that. I will update this comment.

gcaragea commented on src/backend/codegen/include/codegen/codegen_manager.h:
```diff
@@ -0,0 +1,113 @@
+   *
+   * @note Manager manages the memory of enrolled generator.
+   *
+   * @param funcLifespan Based on life span corresponding CodeGen_Utils will be used to generate
```
Comment doesn't sound right. 
"Life span of the enrolled generator. Based on the corresponding life span of the CodeGen_Utils" 
?

karthijrk replied:
yes, Let me correct it as 
"Life span of the enrolling generator. Based on life span,
corresponding CodegenUtils will be used for code generation"

gcaragea commented on src/backend/codegen/include/codegen/codegen_manager.h:
```diff
@@ -0,0 +1,113 @@
+                           CodeGenInterface* generator);
+
+  /**
+   * @brief Make all enrolled generators to generate code.
```
Make -> Request

karthijrk replied:
👍

gcaragea commented on src/backend/codegen/include/codegen/codegen_manager.h:
```diff
@@ -0,0 +1,113 @@
+
+  /**
+   * @brief Compile all the generated functions. On success, caller gets
+   * 		to call the generated method.
```
On success, a pointer to the generated method becomes available to the caller.

karthijrk replied:
👍

gcaragea commented:
I agree with Nikos that we should open a PR if possible only with the renaming changes. That way the overall PR will not look as overwhelming.

gcaragea commented on src/backend/codegen/slot_deform_tuple_codegen.cc:
```diff
@@ -0,0 +1,106 @@
+      llvm_regular_function, forwarded_args);
+
+  // Return the result of the call, or void if the function returns void.
+  if (std::is_same<codegen_utils_detail::FunctionTypeUnpacker<
```
Do we need this here? We should know what type our regular slot_deform_tuple returns. 

It's a nice trick, but not sure if we should complicate things here.

karthijrk replied:
Yes, good catch. I agree with you. We forget to remove this when we refactor it.

gcaragea commented:
This is a lot of very good work! Good job!

Other than a few small asks, mostly regarding comments, I think it looks good. I didn't review the testing framework changes, I'll try to have Shreedhar give me a tour of those changes. 

:+1:

karthijrk commented:
I think I addressed all the comments made here. Thanks a lot guys for spending time. It got better than how it was initially. Based on the comment here, I opened internally two pull request one for rename and another contain actual changes.
1. PR with rename https://github.com/foyzur/gpdb/pull/2
2. PR with actual framework change - https://github.com/foyzur/gpdb/pull/3

karthijrk commented:
Thanks again for the review. I am closing this and as we planned will open in master soon.
